src/migrator/utils.ts

```diff
@@ -23,12 +24,137 @@ export type CategorizedMigrations = {
+export type AsyncMigrationFn = (
+	tx: Tx,
+	options: { batchSize: number },
+	sbvrUtils: SbvrUtils,
+) => Resolvable<Result>;
+
+type AddFn<T extends {}, x extends 'sync' | 'async'> = T & {
+	syncFn: MigrationFn;
+	asyncFn: AsyncMigrationFn;
+} & {
+	[key in `${x}Sql`]?: undefined;
+};
+type AddSql<T extends {}, x extends 'sync' | 'async'> = T & {
+	[key in `${x}Fn`]?: undefined;
+} & {
+	[key in `${x}Sql`]: string;
+};
+
+type BaseAsyncMigration = {
+	type?: 'async';
+	delayMS?: number | undefined;
+	backoffDelayMS?: number | undefined;
+	errorThreshold?: number | undefined;
+	asyncBatchSize?: number | undefined;
+	finalize?: boolean | undefined;
+};
+export type AsyncMigration =
+	| AddFn<BaseAsyncMigration, 'async' | 'sync'>
+	| AddSql<BaseAsyncMigration, 'async' | 'sync'>
+	| AddFn<AddSql<BaseAsyncMigration, 'async'>, 'sync'>
+	| AddFn<AddSql<BaseAsyncMigration, 'sync'>, 'async'>;
 
+export function isAsyncMigration(
+	migration: AsyncMigration | RunnableMigrations,
+): migration is AsyncMigration {
+	return (migration as AsyncMigration).type === MigrationCategories.async;
+}
+
+export function areCategorizedMigrations(
+	migrations: Migrations,
+): migrations is CategorizedMigrations {
+	return Object.keys(migrations).some((key) => key in MigrationCategories);
+}
+
+export type Migration = string | MigrationFn | AsyncMigration;
 export class MigrationError extends TypedError {}
 
+export type MigrationStatus = {
+	migration_key: string;
+	start_time: Date;
+	last_run_time: Date;
+	run_counter: number;
+	migrated_rows: number;
+	error_counter: number;
+	converged_time: Date | undefined;
+	last_error_message: string | undefined;
+	is_backoff: boolean;
+};
+
+export const getRunnableAsyncMigrations = (
+	migrations: Migrations,
+): RunnableAsyncMigrations => {
+	let runnableMigrations: RunnableAsyncMigrations = {};
+	for (const [key, migration] of Object.entries(
+		migrations[MigrationCategories.async],
+	)) {
+		if (isAsyncMigration(migration)) {
+			runnableMigrations = {
+				...runnableMigrations,
+				[key]: migration,
+			};
+		}
+	}
+	return runnableMigrations;
+};
+
+// migration loader should either get migrations from model
+// or from the filepath
+export const getRunnableSyncMigrations = (
+	migrations: Migrations,
+): RunnableMigrations => {
+	if (areCategorizedMigrations(migrations)) {
+		let runnableMigrations: RunnableMigrations = {};
+		Object.entries(migrations).map(([category, categroyMigrations]) => {
+			if (category in MigrationCategories) {
+				Object.entries(categroyMigrations as Migrations).map(
+					([key, migration]) => {
+						if (isAsyncMigration(migration)) {
+							if (migration.finalize) {
+								if (migration.syncFn) {
+									runnableMigrations = {
+										...runnableMigrations,
+										[key]: migration.syncFn,
+									};
+								} else if (migration.syncSql) {
+									runnableMigrations = {
+										...runnableMigrations,
+										[key]: migration.syncSql,
+									};
+								}
```

Comment:
You should be able to switch this to
```suggestion
								runnableMigrations = {
									...runnableMigrations,
									[key]: migration.syncFn ?? migration.syncSql,
								};
```

Reply:
oh thx a lot!